runtime/src/accounts_db.rs

```diff
@@ -989,11 +1001,28 @@ impl AccountsDB {
     }
 
     fn shrink_stale_slot(&self, candidates: &mut MutexGuard<Vec<Slot>>) -> usize {
-        if let Some(slot) = self.do_next_shrink_slot(candidates) {
-            self.do_shrink_stale_slot(slot)
-        } else {
-            0
+        let mut total = 0;
+        let mut count = 0;
+        let start = Instant::now();
+        let num_roots = self.accounts_index.num_roots();
+        loop {
+            if let Some(slot) = self.do_next_shrink_slot(candidates) {
```

Comment:
I think I noticed at one point, that at startup from a snapshot, the `candidates` set was really large. Might be pertinent to check since we're dependent on going through all the candidates before looping back around

Reply:
That's the point of the check for `num_roots / 10` which should limit it to roughly 10% of the candidates set in a single run.